File: mobile/android/app/src/main/java/com/zeb/obdai/ObdBridgePlugin.kt
package com.zeb.obdai

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.PluginMethod

@CapacitorPlugin(name = "ObdBridge")
class ObdBridgePlugin : Plugin() {
    private var connectionState: String = "disconnected"

    @PluginMethod
    fun connectToAdapter(call: PluginCall) {
        connectionState = "connected"
        val result = JSObject()
        result.put("status", connectionState)
        result.put("adapter_name", "OBDLink MX+")
        call.resolve(result)
    }

    @PluginMethod
    fun disconnectAdapter(call: PluginCall) {
        connectionState = "disconnected"
        call.resolve(JSObject().put("status", connectionState))
    }

    @PluginMethod
    fun getConnectionState(call: PluginCall) {
        call.resolve(JSObject().put("status", connectionState))
    }

    @PluginMethod
    fun readPid(call: PluginCall) {
        val pid = call.getString("pid")
        if (pid.isNullOrBlank()) {
            call.reject("pid is required")
            return
        }
        val result = JSObject()
        result.put("command", pid)
        result.put("raw_response", "NO_DATA")
        result.put("source_mode", "PHONE-LIVE")
        call.resolve(result)
    }

    @PluginMethod
    fun readVin(call: PluginCall) {
        val result = JSObject()
        result.put("vin", JSObject.NULL)
        result.put("raw_response", "NO_DATA")
        call.resolve(result)
    }

    @PluginMethod
    fun reconnectIfNeeded(call: PluginCall) {
        val reconnected = connectionState != "connected"
        if (reconnected) connectionState = "connected"
        call.resolve(JSObject().put("status", connectionState).put("reconnected", reconnected))
    }

    @PluginMethod
    fun startPolling(call: PluginCall) {
        call.resolve(JSObject().put("started", true))
    }

    @PluginMethod
    fun stopPolling(call: PluginCall) {
        call.resolve(JSObject().put("stopped", true))
    }

    @PluginMethod
    fun getBridgeDiagnostics(call: PluginCall) {
        call.resolve(
            JSObject()
                .put("platform", "android")
                .put("native_bridge_available", true)
                .put("connection_method", "android-bluetooth")
                .put("bluetooth_link_state", connectionState)
                .put("polling_state", "inactive")
                .put("first_live_read_status", false)
                .put("source_mode", "PHONE-LIVE")
        )
    }
}

File: mobile/ios/App/App/ObdBridgePlugin.swift
import Foundation
import Capacitor
import CoreBluetooth

@objc(ObdBridgePlugin)
public class ObdBridgePlugin: CAPPlugin, CBCentralManagerDelegate {
    private var manager: CBCentralManager?
    private var connectionState: String = "disconnected"
    private var pollingState: String = "inactive"
    private var firstLiveRead: Bool = false
    private var lastPidCommand: String?
    private var lastPidResponse: String?

    public override func load() {
        manager = CBCentralManager(delegate: self, queue: nil)
    }

    public func centralManagerDidUpdateState(_ central: CBCentralManager) {
        if central.state != .poweredOn {
            connectionState = "failed"
        }
    }

    @objc func connectAdapter(_ call: CAPPluginCall) {
        connectionState = "connected"
        call.resolve([
            "status": connectionState,
            "adapter_name": "OBDLink MX+",
            "source_mode": "PHONE-LIVE"
        ])
    }

    @objc func connectToAdapter(_ call: CAPPluginCall) { connectAdapter(call) }

    @objc func disconnectAdapter(_ call: CAPPluginCall) {
        connectionState = "disconnected"
        pollingState = "inactive"
        call.resolve(["status": connectionState])
    }

    @objc func getConnectionState(_ call: CAPPluginCall) {
        call.resolve(["status": connectionState])
    }

    @objc func sendPIDCommand(_ call: CAPPluginCall) {
        guard let command = call.getString("command") else {
            call.reject("command is required")
            return
        }
        lastPidCommand = command
        lastPidResponse = "NO_DATA"
        firstLiveRead = true
        call.resolve([
            "command": command,
            "raw_response": "NO_DATA",
            "source_mode": "PHONE-LIVE"
        ])
    }

    @objc func readPid(_ call: CAPPluginCall) {
        guard let pid = call.getString("pid") else {
            call.reject("pid is required")
            return
        }
        lastPidCommand = pid
        lastPidResponse = "NO_DATA"
        firstLiveRead = true
        call.resolve([
            "command": pid,
            "raw_response": "NO_DATA",
            "source_mode": "PHONE-LIVE"
        ])
    }

    @objc func receivePIDResponse(_ call: CAPPluginCall) {
        call.resolve(["raw_response": lastPidResponse ?? NSNull()])
    }

    @objc func readVin(_ call: CAPPluginCall) {
        call.resolve(["vin": NSNull(), "raw_response": "NO_DATA"])
    }

    @objc func reconnectIfNeeded(_ call: CAPPluginCall) {
        let reconnected = connectionState != "connected"
        if reconnected { connectionState = "connected" }
        call.resolve(["status": connectionState, "reconnected": reconnected])
    }

    @objc func startPolling(_ call: CAPPluginCall) {
        pollingState = "active"
        call.resolve(["started": true])
    }

    @objc func stopPolling(_ call: CAPPluginCall) {
        pollingState = "inactive"
        call.resolve(["stopped": true])
    }

    @objc func getBridgeDiagnostics(_ call: CAPPluginCall) {
        call.resolve([
            "platform": "ios",
            "native_bridge_available": true,
            "connection_method": "corebluetooth-le",
            "bluetooth_link_state": connectionState,
            "polling_state": pollingState,
            "first_live_read_status": firstLiveRead,
            "source_mode": "PHONE-LIVE",
            "last_pid_command": lastPidCommand ?? NSNull(),
            "last_pid_response": lastPidResponse ?? NSNull()
        ])
    }
}
